src/kubernetes.cc

```diff
@@ -1042,27 +1043,47 @@ json::value KubernetesReader::FindTopLevelOwner(
 #endif
   const json::Object* metadata = obj->Get<json::Object>("metadata");
 #ifdef VERBOSE
-  LOG(DEBUG) << "FindTopLevelOwner: metadata is " << *metadata;
+  LOG(DEBUG) << "FindTopLevelController: metadata is " << *metadata;
 #endif
   if (!metadata->Has("ownerReferences")) {
 #ifdef VERBOSE
-    LOG(DEBUG) << "FindTopLevelOwner: no owner references in " << *metadata;
+    LOG(DEBUG) << "FindTopLevelController: no owner references in "
+               << *metadata;
 #endif
     return object;
   }
   const json::Array* refs = metadata->Get<json::Array>("ownerReferences");
 #ifdef VERBOSE
-  LOG(DEBUG) << "FindTopLevelOwner: refs is " << *refs;
+  LOG(DEBUG) << "FindTopLevelController: refs is " << *refs;
 #endif
-  if (refs->size() > 1) {
-    LOG(WARNING) << "Found multiple owner references for " << *obj
+  std::vector<const json::Object*> controller_refs;
+  for (const json::value& ref : *refs) {
+    const json::Object* ref_obj = ref->As<json::Object>();
+    if (ref_obj->Has("controller") &&
+        ref_obj->Get<json::Boolean>("controller")) {
+      controller_refs.emplace_back(ref_obj);
+    }
+  }
+#ifdef VERBOSE
+  LOG(DEBUG) << "FindTopLevelController: contoller_refs is "
+             << controller_refs;
+#endif
+  if (controller_refs.size() == 0) {
+#ifdef VERBOSE
+    LOG(DEBUG) << "FindTopLevelController: no controller owner references in "
+               << *metadata;
+#endif
+    return object;
+  }
+  if (controller_refs.size() > 1) {
+    LOG(WARNING) << "Found multiple controller references for " << *obj
                  << " picking the first one arbitrarily.";
   }
-  const json::value& ref = (*refs)[0];
+  const json::Object controller_ref = *controller_refs[0];
```

Comment:
The comment about not copying JSON values applies here too. Just pass the pointers around.

Reply:
this comment is no longer applicable.